src/com/xilinx/rapidwright/edif/EDIFPortInstList.java

```diff
@@ -0,0 +1,87 @@
+
+    private static final long serialVersionUID = 8718591209309655922L;
+    
+    public static final EDIFPortInstList EMPTY = new EDIFPortInstList();
+    
+    @Override
+    public boolean add(EDIFPortInst e) {
+        int insertionPoint = binarySearch(e.getCellInst(), e.getName());
+        // Do not allow duplicates
+        if(insertionPoint >= 0) {
+            return false;
+        }
+        super.add(insertionPoint >= 0 ? insertionPoint : ~insertionPoint, e);
+        return true;
+    }
+    
+    public EDIFPortInst get(EDIFCellInst i, String name) {
+        int index = binarySearch(i, name);
+        if(index < 0) return null;
+        return get(index);
+    }
+    
+    public EDIFPortInst remove(EDIFPortInst e) {
+        return remove(e.getCellInst(), e.getName());
+    }
+
+    public EDIFPortInst remove(EDIFCellInst inst, String portInstName) {
+        int index = binarySearch(inst, portInstName);
+        if(index < 0) return null;
+        return super.remove(index);
+    }
+    
+    private int binarySearch(EDIFCellInst inst, String portInstName) {
+        String instName = inst == null ? null : inst.getName();
+        int left = 0;
+        int right = size()-1;
+        while(left <= right) {
+            int pivot = (left + right) >>> 1;
+            int result = compare(get(pivot), instName, portInstName);
+            if(result < 0) {
+                left = pivot + 1;
+            } else if (result > 0) {
+                right = pivot - 1;
+            } else {
+                return pivot;
+            }
+        }
+        return ~left;
+    }
+    
+    /**
+     * Performs a 'compareTo' operation without having to create disposable Strings or EDIFPortInst 
+     * objects.  Performs the same operation as 'left.getFullName().compareTo(right.getFullName())'
+     *  where right is the EDIFPortInst represented by rightInstName and rightPortInstName.  
+     * @param left This is the existing EDIFPortInst within the lists that is being compared
+     * @param rightInstName This is the cell instance name {@link EDIFCellInst#getName()} of the 
+     * considered port instance to compare left against.  
+     * @param rightPortInstName This is the port instance name {@link EDIFPortInst#getName()} of the
+     * considered port instance to compare left against.
+     * @return 0 if the left and corresponding right Strings are equal.  A number less than 0 if
+     * left is lexicographically before right, or a number greater than 0 if left is after right. 
+     */
+    private int compare(EDIFPortInst left, String rightInstName, String rightPortInstName) {
+        if(left.getCellInst() == null) {
+            if(rightInstName == null) {
+                return left.getName().compareTo(rightPortInstName);
+            }
+            int compare = left.getName().compareTo(rightInstName);
+            return compare == 0 ? (-1*rightPortInstName.length()) : compare;
```

Comment:
```suggestion
            // left is a top-level port inst, right is not. Compare left's port inst name with the right's inst name
            int compare = left.getName().compareTo(rightInstName);
            // If the two happen to be equal, then left's full name is the prefix of right's full name and so should go before, with the remaining length the number of characters that differ
            return compare == 0 ? (-1*rightPortInstName.length()) : compare;
```

Also, do you need to account for the separator between inst name and port inst name that would otherwise appear when doing `getFullName()`?
After thinking about it some more, I think not (because this situation only occurs when comparing a top-level port inst with a non-top-level inst) unless the return value being off-by-one is critically important to the performance of the sorting algorithm?

Reply:
No, you are correct.  Since we are handling the compare strictly outside of the `getFullName()` scenario, it never becomes an issue.